src/async_fuse/memfs/s3_metadata.rs

```diff
@@ -365,69 +350,20 @@ impl<S: S3BackEnd + Sync + Send + 'static> MetaData for S3MetaData<S> {
             let mut inode = self.get_inode_from_txn(txn.as_mut(), ino).await?;
 
             let origin_attr = inode.get_attr();
-            let applied_attr = if SFlag::S_IFREG == origin_attr.kind {
-                self.apply_local_mtime_and_size(origin_attr)
-            } else {
-                origin_attr
-            };
-
-            let dirty_attr_for_reply = match applied_attr.setattr_precheck(
-                param,
-                context.user_id,
-                context.group_id,
-            )? {
-                Some(mut dirty_attr) => {
-                    // This is to be inserted in KV, whose `mtime` and `size` should be reset to
-                    // origin, because changes of this field are invisible to other
-                    // nodes until flush.
-                    let mut dirty_attr_for_kv = dirty_attr;
-
-                    // There are `mtime` and `size` caches for regular files. Check the change of
-                    // these fields, and set them to the cache.
-                    if SFlag::S_IFREG == applied_attr.kind {
-                        if (dirty_attr.mtime, dirty_attr.size)
-                            != (applied_attr.mtime, applied_attr.size)
-                        {
-                            self.local_mtime_and_size
-                                .insert(ino, (dirty_attr.mtime, dirty_attr.size));
-                        }
-                        // Reset `mtime` and `size` to origin, because changes of these fields is
-                        // invisible to other nodes until flush.
-                        dirty_attr_for_kv.mtime = origin_attr.mtime;
-                        dirty_attr_for_kv.size = origin_attr.size;
-
-                        // The file also needs to be truncated, if the new size is shorter.
-                        if dirty_attr.size < applied_attr.size {
-                            let new_mtime = self
-                                .storage
-                                .truncate(
-                                    ino,
-                                    applied_attr.size.cast(),
-                                    dirty_attr.size.cast(),
-                                    applied_attr.mtime,
-                                )
-                                .await;
-                            self.local_mtime_and_size
-                                .insert(ino, (new_mtime, dirty_attr.size));
-                            dirty_attr.mtime = new_mtime;
-                        }
+            let applied_attr = origin_attr;
+
+            let dirty_attr_for_reply =
+                match applied_attr.setattr_precheck(param, context.user_id, context.group_id)? {
+                    Some(dirty_attr) => {
+                        let dirty_attr_for_kv = dirty_attr;
+                        inode.set_attr(dirty_attr_for_kv);
+                        dirty_attr
                     }
```

Comment:
Why do you remove the calling of `truncate`?

Reply:
Storage related attributes (`mtime` and `size`) will be committed immediately when `setattr` is called, is it right?

Reply:
And, how about `open_files`? Why don't you maintain it in this function?

Reply:
open-to-close only cares about IO operations (expect the `getattr()` ). Other opeartions just directly use the kv.

Reply:
> open-to-close only cares about IO operations (expect the `getattr()` ). Other opeartions just directly use the kv.

So, we must call `truncate` here because it's an IO operation. And `mtime` and `size` should be updated to `open_files` for the following `read` and `write` calls.

Reply:
fixed.

Reply:
So the `setattr` will never interact with `open_files`, would it?

Reply:
Yes